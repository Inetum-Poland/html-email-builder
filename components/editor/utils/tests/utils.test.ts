import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import type { Component, Editor } from "grapesjs";
import grapesJS from "grapesjs";
import grapesJSMJML from "grapesjs-mjml";
import {
  clear,
  exportProject,
  getMjBody,
  getMJML,
  isValidProjectFile,
  selectChildren,
} from "@utils";
import { Format } from "@types";

let editor: Editor;

beforeEach(async (done: () => void) => {
  editor = grapesJS.init({
    container: "#editor",
    plugins: [grapesJSMJML],
  });

  editor.getModel().loadOnStart();
  editor.on("change:readyLoad", done);
});

afterEach(() => {
  editor.destroy();
  vi.clearAllMocks();
});

describe("getMjBody", () => {
  test("returns the mj-body component", () => {
    editor.addComponents("<mjml><mj-body /></mjml>");
    expect(getMjBody(editor)?.attributes.tagName).toBe("mj-body");
  });
});

describe("selectChildren", () => {
  test("selects child components", () => {
    editor.addComponents(
      "<mjml><mj-body><mj-button>test</mj-button></mj-body></mjml>"
    );

    const mjbody = getMjBody(editor);
    selectChildren(editor, mjbody as Component);
    const selectedChildren = editor
      .getSelectedAll()
      .map(({ tagName }: { tagName: string }) => tagName);

    expect(selectedChildren).toStrictEqual(["mj-button"]);
  });
});

describe("getHTML", () => {
  test.todo("returns the HTML content", () => {
  });
});

describe("getMJML", () => {
  test("returns the MJML content", () => {
    const content = "<mjml><mj-body><mj-button>test</mj-button></mj-body></mjml>";
    editor.addComponents(content);
    expect(getMJML(editor)).toBe(content);
  });
});

describe("getJSONProjectData", () => {
  test.todo("returns the JSON project data", () => {
  });
});

describe.skip("exportProject", () => {
  test("downloads the specified HTML data", () => {
    const date = new Date("2024-11-26T16:34:55Z");
    vi.useFakeTimers();
    vi.setSystemTime(date);

    const link = { click: vi.fn() } as any;
    vi.spyOn(document, "createElement").mockImplementation(() => link);

    exportProject(Format.HTML, "test");

    expect(link.download).toEqual("2024-11-26-HTML.html");
    expect(link.href).toContain("blob:nodedata:");
    vi.useRealTimers();
  });

  test("downloads the specified MJML data", () => {
    const date = new Date("2024-11-26T16:34:55Z");
    vi.useFakeTimers();
    vi.setSystemTime(date);

    const link = { click: vi.fn() } as any;
    vi.spyOn(document, "createElement").mockImplementation(() => link);

    exportProject(Format.MJML, "test");

    expect(link.download).toEqual("2024-11-26-MJML.txt");
    expect(link.href).toContain("blob:nodedata:");

    vi.useRealTimers();
  });

  test("downloads the specified JSON data", () => {
    const date = new Date("2024-11-26T16:34:55Z");
    vi.useFakeTimers();
    vi.setSystemTime(date);

    const link = { click: vi.fn() } as any;
    vi.spyOn(document, "createElement").mockImplementation(() => link);

    exportProject(Format.JSON, "test");

    expect(link.download).toEqual("2024-11-26-JSON.json");
    expect(link.href).toContain("blob:nodedata:");

    vi.useRealTimers();
  });
});

describe("copyRichText", () => {
  test.todo("adds the rich text to clipboard", async () => {
  });
});

describe("resize", () => {
  test.todo("resizes the specified image", () => {
  });
});

describe("uploadImage", () => {
  test.todo("uploads the specified image", () => {
  });
});

describe("restore", () => {
  test.todo("restores the project state", () => {
  });
});

describe("clear", () => {
  test("clears the project state", () => {
    const content = "<mjml><mj-body><mj-button>test</mj-button></mj-body></mjml>";
    editor.addComponents(content);
    expect(getMJML(editor)).toBe(content);
    clear(editor);
    expect(getMJML(editor)).toBe("<mjml><mj-body></mj-body></mjml>");
  });
});

describe("upload", () => {
  test.todo("executes callback with the uploaded data", () => {
  });
});

describe("isValidProjectFile", () => {
  test("is true when file has all required properties", () => {
    const data = {
      assets: [],
      styles: [],
      pages: [],
      symbols: [],
      dataSources: []
    };

    expect(isValidProjectFile(data)).toBe(true);
  });

  test("is false when file does not have all required properties", () => {
    const data = {
      assets: [],
      styles: [],
      pages: [],
    };

    expect(isValidProjectFile(data)).toBe(false);
  });

  test("is true when file has extra properties", () => {
    const data = {
      assets: [],
      styles: [],
      pages: [],
      symbols: [],
      dataSources: [],
      extra: [],
      comment: "comment",
    };

    expect(isValidProjectFile(data)).toBe(true);
  });
});
