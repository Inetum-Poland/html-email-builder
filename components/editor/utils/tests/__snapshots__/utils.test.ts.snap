// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getHTML > returns the HTML content 1`] = `
"<!doctype html>
<html xmlns="http://www.w3.org/1999/xhtml" xmlns:v="urn:schemas-microsoft-com:vml" xmlns:o="urn:schemas-microsoft-com:office:office">
  <head>
    <title></title>
    <!--[if !mso]><!-->
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <!--<![endif]-->
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type="text/css">
      #outlook a { padding:0; }
      body { margin:0;padding:0;-webkit-text-size-adjust:100%;-ms-text-size-adjust:100%; }
      table, td { border-collapse:collapse;mso-table-lspace:0pt;mso-table-rspace:0pt; }
      img { border:0;height:auto;line-height:100%; outline:none;text-decoration:none;-ms-interpolation-mode:bicubic; }
      p { display:block;margin:13px 0; }
    </style>
    <!--[if mso]>
    <noscript>
    <xml>
    <o:OfficeDocumentSettings>
      <o:AllowPNG/>
      <o:PixelsPerInch>96</o:PixelsPerInch>
    </o:OfficeDocumentSettings>
    </xml>
    </noscript>
    <![endif]-->
    <!--[if lte mso 11]>
    <style type="text/css">
      .mj-outlook-group-fix { width:100% !important; }
    </style>
    <![endif]-->
    
      <!--[if !mso]><!-->
        <link href="https://fonts.googleapis.com/css?family=Ubuntu:300,400,500,700" rel="stylesheet" type="text/css">
        <style type="text/css">
          @import url(https://fonts.googleapis.com/css?family=Ubuntu:300,400,500,700);
        </style>
      <!--<![endif]-->

    
    
    <style type="text/css">
    
    
    </style>
    <style type="text/css">
    
    </style>
    
  </head>
  <body style="word-spacing:normal;">
    
    
      <div
         style=""
      >
        
      <table
         border="0" cellpadding="0" cellspacing="0" role="presentation" style="border-collapse:separate;line-height:100%;"
      >
        <tbody>
          <tr>
            <td
               align="center" bgcolor="#414141" role="presentation" style="border:none;border-radius:3px;cursor:auto;mso-padding-alt:10px 25px;background:#414141;" valign="middle"
            >
              <p
                 style="display:inline-block;background:#414141;color:#ffffff;font-family:Ubuntu, Helvetica, Arial, sans-serif;font-size:13px;font-weight:normal;line-height:120%;margin:0;text-decoration:none;text-transform:none;padding:10px 25px;mso-padding-alt:0px;border-radius:3px;"
              >
                test
              </p>
            </td>
          </tr>
        </tbody>
      </table>
    
      </div>
    
  </body>
</html>
  "
`;

exports[`getJSONProjectData > returns the JSON project data 1`] = `
{
  "assets": [],
  "dataSources": [],
  "pages": [
    {
      "frames": [
        {
          "component": {
            "components": [
              {
                "components": [
                  {
                    "attributes": {
                      "style": "width:600px;",
                      "width": "600px",
                    },
                    "components": [
                      {
                        "attributes": {
                          "align": "center",
                          "background-color": "#414141",
                          "border-radius": "3px",
                          "color": "#ffffff",
                          "font-size": "13px",
                          "font-weight": "400",
                          "padding-bottom": "10px",
                          "padding-left": "25px",
                          "padding-right": "25px",
                          "padding-top": "10px",
                          "style": "background-color:#414141;border-radius:3px;font-size:13px;font-weight:400;color:#ffffff;vertical-align:middle;padding-top:10px;padding-bottom:10px;padding-right:25px;padding-left:25px;align:center;",
                          "vertical-align": "middle",
                        },
                        "components": [
                          {
                            "content": "test",
                            "type": "textnode",
                          },
                        ],
                        "style": {
                          "align": "center",
                          "background-color": "#414141",
                          "border-radius": "3px",
                          "color": "#ffffff",
                          "font-size": "13px",
                          "font-weight": "400",
                          "padding-bottom": "10px",
                          "padding-left": "25px",
                          "padding-right": "25px",
                          "padding-top": "10px",
                          "vertical-align": "middle",
                        },
                        "tagName": "mj-button",
                        "type": "mj-button",
                      },
                    ],
                    "style": {
                      "width": "600px",
                    },
                    "tagName": "mj-body",
                    "type": "mj-body",
                  },
                ],
                "tagName": "mjml",
                "type": "mjml",
              },
            ],
            "docEl": {
              "tagName": "html",
            },
            "head": {
              "type": "head",
            },
            "stylable": [
              "background",
              "background-color",
              "background-image",
              "background-repeat",
              "background-attachment",
              "background-position",
              "background-size",
            ],
            "type": "wrapper",
          },
          "id": Any<String>,
        },
      ],
      "id": Any<String>,
      "type": "main",
    },
  ],
  "styles": [],
  "symbols": [],
}
`;
